pkg/pillar/types/sriovtypes.go

```diff
@@ -0,0 +1,225 @@
+		return nil
+	}
+	pf, err := netlink.LinkByName(vf.Iface)
+	if err != nil {
+		return fmt.Errorf("Failed to find physical function %s: %v", vf.Iface, err)
+	}
+
+	if err = netlink.LinkSetVfVlan(pf, int(vf.Index), int(vf.VlanId)); err != nil {
+		return fmt.Errorf("Failed to set vf %d vlan: %v", vf.Index, err)
+	}
+	return nil
+}
+
+// VFList is list of VF for given PF (Eth device)
+type VFList struct {
+	Count uint8
+	Data  []EthVF
+}
+
+// UsedBy finds VF used by app UUID
+func (l *VFList) UsedBy(appUuid uuid.UUID) *EthVF {
+	for _, vf := range l.Data {
+		if vf.UsedByUUID == appUuid {
+			return &vf
+		}
+	}
+	return nil
+}
+
+// UsedById finds VF used by VF id
+func (l *VFList) UsedById(id uint8) *EthVF {
+	for _, vf := range l.Data {
+		if vf.Index == id {
+			return &vf
+		}
+	}
+	return nil
+}
+
+func (vfs *VFList) setupVlans() error {
+	for _, vf := range vfs.Data {
+		if err := vf.setupVlan(); err != nil {
+			return fmt.Errorf("failed to setup link %v", err)
+		}
+	}
+	return nil
+}
+
+// NicVFs is structure to store VFs for interfaces
+type NicVFs struct {
+	IfaceVf map[string]*VFList
+}
+
+// VfInfo read VF information from filesystem for each PF in folders
+func (vfs *NicVFs) VfInfo(nicDir string) error {
+	files, err := ioutil.ReadDir(nicDir)
+
+	if err != nil {
+		return err
+	}
+
+	for _, iface := range files {
+		ethVfs, err := GetVf(nicDir + iface.Name())
+		if err == nil && ethVfs != nil {
+			vfs.IfaceVf[iface.Name()] = ethVfs
+		}
+	}
+	return nil
+}
+
+// AssignVf sets app UUID for interface
+func (vfs *NicVFs) AssignVf(id uuid.UUID, iface string) error {
+	if val, ok := vfs.IfaceVf[iface]; !ok {
+		return fmt.Errorf("iface %s is not in NicVFs map", iface)
+	} else {
+		for _, vf := range val.Data {
+			if vf.UsedByUUID != uuid.Nil {
+				vf.UsedByUUID = id
+			}
+		}
+	}
+	return nil
+}
+
+// FromIoAdapterList updates VF information from ZedAgent
+func (pfs *NicVFs) FromIoAdapterList(ioAdapterList []IoAdapter) error {
+	for _, a := range ioAdapterList {
+		if a.Type == IoNetEthVF || a.Type == IoNetEthPF {
+			if vfs, ok := pfs.IfaceVf[a.Name]; ok {
+				if vf := vfs.UsedById(a.EthVf.Index); vf != nil {
+					vf.fromAdapter(a.EthVf)
+					if a.EthVf.Mac != nil {
+						vf.setupLink()
+					}
+					if a.EthVf.VlanId != 0 {
+						vf.setupVlan()
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// GetVf retrieve information about VFs for NIC given
+func GetVf(nicPath string) (*VFList, error) {
+	var res []EthVF
+	virtfnRe := regexp.MustCompile("(virtfn)(\\d{1,})")
+	pciBdfRe := regexp.MustCompile("[0-9a-f]{4}:[0-9a-f]{2,4}:[0-9a-f]{2}\\.[0-9a-f]")
+	devPath := nicPath + "/device/"
+
+	devInfo, err := os.Stat(devPath)
+	if err != nil {
+		return nil, fmt.Errorf("vfInfo failed. Cannot obtain get %s path info. Error: %s", devPath, err)
+	}
+	physfnInfo, err := os.Lstat(devPath + "/physfn")
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("Cannot obtain get %s path info. Error: %s", devPath+"/physfn", err)
```

Comment:
Don't you get these errors all the time; from what I can tell you call GetVf on everything in /sys/class/net/